File: solve/22859.py
# https://www.acmicpc.net/problem/22859

import sys
sys.setrecursionlimit(11111)
doc = sys.stdin.readline().strip()
pos = 0

def moveUntil(ch):
    global pos

    buf = []
    while pos < len(doc) and doc[pos] != ch:
        buf.append(doc[pos])
        pos += 1
    return "".join(buf)

def readTag():
    global pos

    pos += 1 # <
    if doc[pos] == '/':
        pos += 1
        tagName = moveUntil('>')
        pos += 1 # >
        return { "name": tagName, "isOpening": False }
    elif pos + 3 < len(doc) and \
            doc[pos + 0] == 'd' and \
            doc[pos + 1] == 'i' and \
            doc[pos + 2] == 'v' and \
            doc[pos + 3] == ' ':
        moveUntil('"')
        pos += 1
        title = moveUntil('"')
        moveUntil('>')
        pos += 1
        return { "name": "div", "isOpening": True, "title": title }
    else:
        tagName = moveUntil('>')
        pos += 1 # >
        return { "name": tagName, "isOpening": True }

def readParagraph():
    p = ""

    while pos < len(doc):
        token = moveUntil('<')
        p += token

        tag = readTag()
        if tag["name"] == 'p' and not tag["isOpening"]:
            break

    p = ' '.join(p.split())
    p = p.strip()

    return p

while pos < len(doc):
    tag = readTag()
    if tag["name"] == "main":
        continue
    elif tag["name"] == "div" and tag["isOpening"]:
        print("title : " + tag["title"])
    elif tag["name"] == "p" and tag["isOpening"]:
        p = readParagraph()
        print(p)

File: solve/22859.c
// https://www.acmicpc.net/problem/22859

#include <stdio.h>

char doc[1111111];
char buf[1111111];
int pos;

int moveUntil(char ch)
{
    int d = 0;
    while (doc[pos + d] != ch)
    {
        d++;
    }
    return d;
}

int putUntil(char ch)
{
    int d = 0;
    while (doc[pos + d] != ch)
    {
        putchar(doc[pos + d]);
        d++;
    }
    return d;
}

int main()
{
    fread(doc, 1, sizeof(doc), stdin);
    //fgets(doc, sizeof(doc), stdin);
    pos = 6; // <main>

    while (doc[pos + 1] == 'd') // <div
    {
        pos += moveUntil('"') + 1; // title="
        printf("title : ");
        pos += putUntil('"') + 2; // ">
        putchar('\n');
        while (doc[pos + 1] == 'p')
        {
            pos += 3; // <p>

            int bufLen = 0;
            int spaced = 1;
            while (1) // <p> 부터 </p> 만날때까지 반복
            {
                for (; doc[pos] != '<'; pos++)
                {
                    if (doc[pos] == ' ')
                    {
                        if (!spaced)
                        {
                            buf[bufLen++] = ' ';
                            spaced = 1;
                        }
                    }
                    else
                    {
                        buf[bufLen++] = doc[pos];
                        spaced = 0;
                    }
                }

                if (doc[pos + 1] == '/' &&
                    doc[pos + 2] == 'p' &&
                    doc[pos + 3] == '>')
                {
                    pos += 4; // </p>
                    break;
                }
                else
                {
                    pos += moveUntil('>') + 1;
                }
            }

            while (bufLen >= 0 && buf[bufLen - 1] == ' ')
                bufLen--;
            buf[bufLen] = '\0';
            puts(buf);
        }
        pos += 6; // </div>
    }
    return 0;
}